core/src/main/java/org/elasticsearch/common/settings/Setting.java

```diff
@@ -0,0 +1,410 @@
+            return new Updater(consumer, logger, settings, accept);
+        } else {
+            throw new IllegalStateException("setting [" + getKey() + "] is not dynamic");
+        }
+    }
+
+    /**
+     * this is used for settings that depend on each other... see {@link org.elasticsearch.common.settings.AbstractScopedSettings#addSettingsUpdateConsumer(Setting, Setting, BiConsumer)} and it's
+     * usage for details.
+     */
+    static <A, B> AbstractScopedSettings.SettingUpdater compoundUpdater(final BiConsumer<A,B> consumer, final Setting<A> aSettting, final Setting<B> bSetting, ESLogger logger, Settings settings) {
+        final AtomicReference<A> aRef = new AtomicReference<>();
+        final AtomicReference<B> bRef = new AtomicReference<>();
+        final AbstractScopedSettings.SettingUpdater aSettingUpdater = aSettting.newUpdater(aRef::set, logger, settings);
+        final AbstractScopedSettings.SettingUpdater bSettingUpdater = bSetting.newUpdater(bRef::set, logger, settings);
+        return new AbstractScopedSettings.SettingUpdater() {
+            boolean aHasChanged = false;
+            boolean bHasChanged = false;
+            @Override
+            public boolean prepareApply(Settings settings) {
+                aHasChanged = aSettingUpdater.prepareApply(settings);
+                bHasChanged = bSettingUpdater.prepareApply(settings);
+                return aHasChanged || bHasChanged;
+            }
+
+            @Override
+            public void apply() {
+                aSettingUpdater.apply();
+                bSettingUpdater.apply();
+                if (aHasChanged || bHasChanged) {
+                    consumer.accept(aRef.get(), bRef.get());
+                }
+            }
+
+            @Override
+            public void rollback() {
+                try {
+                    aRef.set(null);
+                    aSettingUpdater.rollback();
+                } finally {
+                    bRef.set(null);
+                    bSettingUpdater.rollback();
+                }
+            }
+
+            @Override
+            public String toString() {
+                return "CompoundUpdater for: " + aSettingUpdater + " and " + bSettingUpdater;
+            }
+        };
+    }
+
+
+    private class Updater implements AbstractScopedSettings.SettingUpdater {
+        private final Consumer<T> consumer;
+        private final ESLogger logger;
+        private final Consumer<T> accept;
+        private String value;
+        private boolean commitPending;
+        private String pendingValue;
+        private T valueInstance;
+
+        public Updater(Consumer<T> consumer, ESLogger logger, Settings settings,  Consumer<T> accept) {
+            this.consumer = consumer;
+            this.logger = logger;
+            value = getRaw(settings);
+            this.accept = accept;
+        }
+
+
+        public boolean prepareApply(Settings settings) {
+            String newValue = settings.get(key);
```

Comment:
do we need this extra get? it seems `newValue = getRaw(settings);` will give you this value if it's non-null. It does save on a default value calculation, but if we care about this, I'm not sure this is the right place?